system/security/securesocket/securesocket.cpp

```diff
@@ -1142,6 +1142,161 @@ const char* strtok__(const char* s, const char* d, StringBuffer& tok)
     return s;
 }
 
+static bool use_PrivateKey_PEM_buffer(SSL_CTX *ctx, const char *privkeybuf)
+{
+    // single RSA key in buffer
+
+    BIO *cbio = BIO_new_mem_buf((void *)privkeybuf, -1);
+    if (!cbio)
+        return false;
+
+    RSA *rsa = PEM_read_bio_RSAPrivateKey(cbio, NULL, 0, NULL);
+    if (rsa == NULL)
+    {
+        BIO_free(cbio);
```

Comment:
Worth using a wrapper class for a BIO * which calls BIO_free on exit.  It would clean up the code in this and other functions. 
... I see Jake has suggested the same below.    I think the following works (it compiles!), and might be a good pattern to adopt elsewhere:

```
struct BIO_deleter
{
    void operator() ( BIO * bio )
    {
       BIO_free( bio ); 
    }
};
using unique_bioptr = std::unique_ptr<BIO, BIO_deleter>;
...
unique_bioptr cbio(BIO_new_mem_buf((void *)privkeybuf, -1));
``` 
@jakesmith thoughts?

Reply:
std::unique_ptr vs cryptohelper::OwnedEVPBio would be okay too, and more standard.
Only -ve is that std::unique_ptr doesn't implicitly cast to the underlying pointer the way we are used to with our Owned* varieties [i.e. u need to call uptr.get() ]

Either would make the code cleaner/safer.
It would be nice if we consistently use the same pattern though.

Reply:
ok, added for BIO.
Perhaps could also do same for stack of x509 ...

Reply:
yes, I think in almost all cases, it would be cleaner/safer and make the code more robust in general.